File: Algorithm/Python/572. Subtree of Another Tree.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2017-06-16 09:34:18
# @Last modified by:   WuLC
# @Last Modified time: 2017-06-16 09:49:50
# @Email: liangchaowu5@gmail.com


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# recursive
class Solution(object):
    def isSubtree(self, s, t):
        """
        :type s: TreeNode
        :type t: TreeNode
        :rtype: bool
        """
        if t == None and s == None:
            return True
        if t == None or s == None:
            return False
        if s.val == t.val and self.rootSubtree(s, t):
            return True
        return self.isSubtree(s.left, t) or self.isSubtree(s.right, t)
    
    def rootSubtree(self, s, t):
        if t == None and s == None:
            return True
        if t == None or s == None or s.val != t.val:
            return False
        return self.rootSubtree(s.left, t.left) and self.rootSubtree(s.right, t.right)


# same as above, but more graceful
class Solution(object):
    def isSubtree(self, s, t):
        if s == None:
            return False
        if self.isSame(s, t):
            return True
        return self.isSubtree(s.left, t) or self.isSubtree(s.right, t)
    
    
    def isSame(self, s, t):
        if s == None and t == None:
            return True
        if s == None or t == None:
            return False
        if s.val != t.val:
            return False
        return self.isSame(s.left, t.left) and self.isSame(s.right, t.right)

File: Algorithm/Java/572. Subtree of Another Tree.java
/*
 * Created on Sun Feb 11 2018 15:49:33
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

 // recursive
class Solution 
{
    public boolean isSubtree(TreeNode s, TreeNode t) 
    {
        if (t == null) 
            return true;
        else if(s == null) 
            return false;
        else
            return sameTree(s, t) || isSubtree(s.left, t) || isSubtree(s.right, t);
    }
    
    private boolean sameTree(TreeNode t1, TreeNode t2)
    {
        if (t1==null && t2 == null)
            return true;
        else if (t1==null || t2==null)
            return false;
        else
            return t1.val==t2.val && sameTree(t1.left, t2.left) && sameTree(t1.right, t2.right);
    }
}